src/mesh/boundary_info.C

```diff
@@ -1865,9 +1877,154 @@ BoundaryInfo::build_node_list_from_side_list()
 
           // Add each node node on the side with the side's boundary id
           for (unsigned int i=0; i<side->n_nodes(); i++)
-            this->add_node(side->node_ptr(i), pos->second.second);
+            {
+              const boundary_id_type bcid = pos->second.second;
+              this->add_node(side->node_ptr(i), bcid);
+              if (!mesh_is_serial)
+                {
+                  const processor_id_type proc_id =
+                    side->node_ptr(i)->processor_id();
+                  if (proc_id != my_proc_id)
+                    nodes_to_push[proc_id].insert
+                      (std::make_pair(side->node_id(i), bcid));
+                }
+            }
+        }
+    }
+
+  // If we're on a serial mesh then we're done.
+  if (mesh_is_serial)
+    return;
+
+  // Otherwise we need to push ghost node bcids to their owners, then
+  // pull ghost node bcids from their owners.
+  Parallel::MessageTag
+    node_pushes_tag = this->comm().get_unique_tag(31337),
+    node_pulls_tag = this->comm().get_unique_tag(31338),
+    node_responses_tag = this->comm().get_unique_tag(31339);
+
+  std::vector<Parallel::Request> node_push_requests(n_proc-1);
+
+  for (processor_id_type p = 0; p != n_proc; ++p)
+    {
+      if (p == my_proc_id)
+        continue;
+
+      Parallel::Request &request =
+        node_push_requests[p - (p > my_proc_id)];
+
+      this->comm().send
+        (p, nodes_to_push[p], request, node_pushes_tag);
+    }
+
+  for (processor_id_type p = 1; p != n_proc; ++p)
+    {
+      set_type received_nodes;
+
+      this->comm().receive
+        (Parallel::any_source, received_nodes, node_pushes_tag);
+
+      for (set_type::const_iterator it = received_nodes.begin(),
+             end = received_nodes.end(); it != end; ++it)
+        this->add_node(_mesh.node_ptr(it->first), it->second);
+    }
+
+  // At this point we should know all the BCs for our own nodes; now
+  // we need BCs for ghost nodes.
+  //
+  // FIXME - parallel_ghost_sync.h doesn't work here because it
+  // assumes a fixed size datum on each node.
+  std::vector<std::vector<dof_id_type> > node_ids_requested(n_proc);
+
+  // Determine what nodes we need to request
+  for (MeshBase::const_node_iterator it = _mesh.nodes_begin(),
```

Comment:
Right, the `cond` part involves calling `_mesh.nodes_end()` repeatedly, and this is slower than hosting it outside the loop IIRC.  I can try to make a test code and verify that's still the case...